### PR TITLE
[Linux MCU] Ability to use multiple GPIO chips

### PR DESCRIPTION
**Pull request by dianlight**
Extends the implementation and allows you to use /dev/gpiochip from 0 to 8.

This makes it possible to use GPIO also on systems other than RPI where the exposed pins do not always belong to the first chip. (Issue #2916 )

Also as a bonus it allows the use of GPIO expander recognized by the kernel (such as MCP2307 / MCP2308) in a transparent way.

~~The patch consists of both a C part and a change to the make scripts to allow you to set the amount of gpiochips if you compile it on a non-target system.~~

**ATTENTION**: This PR changes the pin names for linux_mcu from `P<gpio>` to `gpiochip<chip>\gpio<gpio>` with short name for first chip,  so for example `P20` becomes `gpiochip0\gpio20` or `gpio20`.

### CODE REVIEW

**Comment by KevinOConnor** [comment]
Thanks.  Some comments:
1. I do not think we should add a Kconfig setting or makefile check for this.  Just assume a large number of gpio chips (eg, 16) and verify that the file exists at runtime.  If the user attempts to use a pin that doesn't exist then raise an error by calling `shutdown()`.
2. Don't use an "odd" number for MAX_GPIO_LINES - use 256 so that the compiler can optimize the math.
3. If the range of gpio pin enumerations is changed then analog.c and hard_pwm.c need to be checked to ensure their enumerations don't conflict.
4. I'd like to avoid creating "yet another alias" for gpio pins.  I recommend using the name "gpio123" for gpiochip0 pins and using the name "gpiochip1/gpio123" for all other gpiochip pins.

@andryblack - do you have any comments on this?

-Kevin

**Comment by andryblack** [comment]
I completely agree with the comments of Kevin.

File: src/linux/gpio.c [hunk 1]
```diff
@@ -19,45 +19,76 @@
...
-    if (gpio_chip_fd <= 0) {
-        gpio_chip_fd = open(CHIP_FILE_NAME,O_RDWR | O_CLOEXEC);
-        if (gpio_chip_fd < 0) {
-            report_errno("open " CHIP_FILE_NAME,-1);
+    if (gpio_chip_fd[chipId] <= 0) {
+        char *chipFilename = malloc(sizeof(CHIP_FILE_NAME)+1);
+        sprintf(chipFilename, CHIP_FILE_NAME, chipId);
+        gpio_chip_fd[chipId] = open(chipFilename,O_RDWR | O_CLOEXEC);
+        if (gpio_chip_fd[chipId] < 0) {
+            char *errorMessage =
```
Comment (andryblack) [inline src/linux/gpio.c:71]: use on-stack or static buffer

**Comment by dianlight** [comment]
> 4\. I'd like to avoid creating "yet another alias" for gpio pins.  I recommend using the name "gpio123" for gpiochip0 pins and using the name "gpiochip1/gpio123" for all other gpiochip pins.

@KevinOConnor the rest is clear but I don't understand the difference of names for gpiochip0. 
Now the gpio are named as P0..n and the change would not be compatible with the current one, so why not standardize and call gpiochip0/gpio0..n?
Also in hard_pwm.c is pwmchip0/pwm0 and not simply pwm0.

L.

**Comment by KevinOConnor** [comment]
> Now the gpio are named as P0..n and the change would not be compatible with the current one, so why not standardize and call gpiochip0/gpio0..n?

I'm expecting the vast majority of users to have just a gpiochip0 (that is, I expect the vast majority of users to be using a Raspberry Pi or similar device).  I think it would be easier to document that the pin name is generally of the form "gpio23" (or perhaps "GPIO23") for these users.

It should be possible to add overlapping enumerations, so you can add mappings for both "gpiochip0/gpioX" and "gpioX" to the same values, if desired.

> Also in hard_pwm.c is pwmchip0/pwm0 and not simply pwm0.

I'm not expecting many users to directly define a pwm pin (it's typically setup by the replicape.py code), so using a more verbose name makes sense here.

Cheers,
-Kevin

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Pull request merged**